Add SD subsystem into Zephyr

danieldegrasse opened the pull request:
# SD Subsystem RFC

## Introduction
I would like to implement an abstraction layer for SD host drivers, similar to what exists in Linux, that abstracts the host controller from the SD protocol. This would place all common SD protocol code into one subsystem, and enable SDIO device support as well as standard memory card support, only requiring vendors to implement drivers for their SDHC peripheral.

## Problem description
Although large portions of the SDMMC standard are common between all SD host controllers, the implementation of SDMMC support in Zephyr is currently split between several vendor-specific drivers. This means that each vendor's driver reimplements the SD host specification, and that new features or bugfixes added to one driver will not be present in the others.

## Proposed change
This change will separate the SD host controller logic (which can be vendor specific) from the SD card specification. SD memory cards and SDIO cards will have subsystem APIs exposed to configure and write to them. SD memory cards can still be used as disk devices via a shim disk driver that relies on the SD memory card subsystem. All vendor specific code will exist at the SD host controller level.

The SD subsystem will handle components of SD card communication defined within the specification, including the initialization and configuration of cards.

For example, the required commands for determining if an SD card supports 1.8V signaling would all be handled by the SD subsystem, but the details of sending the commands themselves to the SD card would be handled by the SD host controller driver and the underlying SD peripheral.

# New APIs

## High level API Diagrams

Below is a block diagram of the proposed SD subsystem implementation:
![SD Subsystem Architecture](https://user-images.githubusercontent.com/34665051/155623078-22c53251-6bbd-49d5-8624-ca0d53be3ec8.png)


For reference, here is a block diagram describing the current implementation of SDMMC drivers in Zephyr:
![Current SDMMC Architecture](https://user-images.githubusercontent.com/34665051/155623111-5bfa8fa7-8808-417d-8e8b-fa84af506658.png)


Note that the specifics of the current implementation may vary at the vendor level. The only portion that is shared between all SDMMC drivers is the api they expose, the disk driver API.

## Common SD API
All consumers of SD devices will start by calling `sd_init`, which will query the card to determine if it is a memory card, or SD I/O device. Once `sd_init` is called, the user can use SDMMC or SDIO specific APIs.
The common API exposes the following functions:
- `sd_init`: initialize an SD card when one is present
- `sd_is_card_present`: check if card is physically present in SD slot

The SD common API is fully implemented and tested. The implementation resides in `subsys/sd/sd.c`, and the api is defined in `include/sd/sd.h`


## SDMMC API
The SDMMC API will be implemented within the SD subsystem. The disk driver will consume the SDMMC subsystem api. The SDMMC api will expose the following functions:
- `sdmmc_write_blocks`: write data blocks to SD card
- `sdmmc_read_blocks`: read data blocks from SD card
- `sdmmc_ioctl`: get IO control information for SD card

The SDMMC subsystem API is fully implemented and tested. The implementation resides in `subsys/sd/sdmmc.c`, and the api is defined in `include/sd/sdmmc.h`

See `tests/subsys/sd/sdmmc` for a project using the SD subsystem API, or `drivers/disk/sdmmc_subsys.c` for a disk driver shim using the SD subsystem.

## SDIO API

**Note: This portion of the API is simply a proposal, and may change when the final implementation is designed**

SDIO support will be implemented within the SD subsystem. Drivers wishing to use SDIO devices can manage them by consuming the api exposed by the SD middleware layer, like the disk driver does for SDMMC devices. The proposed SDIO api will contain the following functions:
- `sdio_io_write`: direct I/O write to SDIO card. First register to write to is provided. If a count of more than one is requested, an extended write will be used to write multiple SDIO registers
- `sdio_io_read`: direct I/O read from SD card. First register to read from is provided. If a count of more than one is requested, an extended read will be used to read from multiple SDIO registers
- `sdio_func_enable`: enable SDIO function. Required before sdio_io_write or sdio_io_read can be used with this function.

Note that the SDIO API is not implemented at this time, because there is no infrastructure within Zephyr that would use SDIO devices, limiting testing possibilities.

## SD Host Controller API
SD host controllers will all have vendor specific implementations, but they will expose a common SDHC api that the SD subsystem will consume (an application may consume this API directly, but there is rarely a reason to).
The SDHC api will expose the following functions:
- `sdhc_hw_reset`: reset the host controller state (do not reset the SD card)
- `sdhc_request`: send a command, and optionally data as well, to the SD card
- `sdhc_set_io`: configure clocks, IO voltage level, and bus width for the peripheral. Can also toggle power to the SD card.
- `sdhc_card_present`: check if a card is present in the slot
- `sdhc_execute_tuning`: execute clock tuning, required when using higher clock speeds for cards
- `sdhc_card_busy`: check if card bus is busy
- `sdhc_get_host_props`: get host properties, including maximum clock frequency, host capabilities, and maximum host current.

The SDHC api is fully implemented and tested. Two host controllers have been implemented:
- NXP's USDHC peripheral, in `drivers/sdhc/imx_usdhc.c`
- A generic SPI mode host controller, in `drivers/sdhc/sdhc_spi.c`

See `drivers/sdhc.h` for the SD host controller API definition.

The SD subsystem consumes the SDHC api, but `tests/drivers/sdhc` contains a simple use of the SDHC api that performs the early initialization steps for an SD device.

# Testing
This RFC introduces two new tests:
- `tests/drivers/sdhc/sdhc_api`, which verifies basic API functionality for SD host controller drivers
- `tests/subsys/sd/sdmmc`, which verifies the SD stack is correctly able to initialize SDMMC cards and read/write to them

The stack was primarily tested on the `mimxrt1064_evk`, but has been verified to work on all RT evaluation kits with a USDHC peripheral. It has also been tested using the `lpcxpresso55s69_cpu0` target, to verify the SD SPI mode host driver.


# Future Work

## SDIO support

This subsystem is intended to lay the groundwork for supporting SDIO devices in Zephyr, but the stack is not implemented. A proposed SDIO api has been defined, but implementation of the SDIO stack is currently hindered by available testing methods, since SDIO devices do not expose a common type of device, like a disk, and a vendor specific SDIO driver would need to be implemented to fully test any SDIO stack implementation. As such, the SDIO api is only a proposal, and would likely not be merged with the initial SDMMC subsystem implementation.

## Dynamic loading

In the initial pass of this implementation, all definitions of SD devices will be statically instantiated from the devicetree. However, SD is by design hot pluggable, and dynamically identifying and enumerating SD devices is a future goal of this work. One potential implementation would be to allow each SDIO/SDMMC driver to register an identifier value for the card (for example, the card identification number for SDMMC devices or the function identification tuple in SDIO devices), and a callback function. When a new SD device is identified, the SD subsystem could compare the identifier values for each registered driver, and call the callback of the first matching driver.

# Detailed RFC

## Devicetree example
The devicetree for a board with an SD memory card present on the `SDHC_0` peripheral and SDIO wifi card present on the `SDHC_1` peripheral might look like this:

```
&sdhc0 {
    status = "okay";
    compatible = "vendor,sdhc";
    mmc: mmc {
        compatible = "zephyr,sdmmc-disk";
        status = "okay";
        label = "SDMMC_0";
    };
};

&sdhc1 {
	status = "okay";
	compatible = "vendor, sdhc";
	wifi: sdio {
		compatible = "vendor,wifi";
		status = "okay";
		label = "WIFI_0";
	};
};
```

In this example, the `zepher,sdmmc-disk` driver will implement the disk driver interface defined in `drivers/disk.h`, and would make calls to the SDMMC portion of the SD subsystem. The SD subsystem would in turn use the SDHC api to manage the SD card. The `vendor,wifi` driver would make calls to the SDIO portion of the subsystem, which would then use the SDHC api to manage the SDIO card. The `vendor,sdhc` driver would be vendor specific code implementing a driver conforming to the SDHC api.

The SD subsystem will leverage the same code where possible to manage SDIO and SDMMC devices, but expose two separate APIs, so that the similarities between the SDIO and SDMMC specification can be handled by the same functions.

## Detailed Initialization Flow

The below diagram is intended as a real example of where each portion of the SD host controller api comes into play during the SD initialization sequence.
the flow roughly follows the command sequence sent during a call to `sd_init`, but is primarily intended to provide a concrete example of how the SDHC api is used.

1. Initialize the host controller
2. Check that the card is present
3. Toggle power to the card to clear any old configuration
4. Request all cards on the bus go idle (this resets card state) (`CMD0`)
5. Query the SD card to determine if it is a high capacity card (`CMD8`)
6. Query the SD card to determine if the card supports 1.8V (`ACMD41`)
7. If the card supports 1.8V reduce signaling voltage to 1.8V (`CMD11`)
8. Get the card's identifier (`CMD2`)
9. Set the card's relative address (`CMD3`)
10. Raise the bus width if the card supports it (`ACMD6`)
11. Select the bus timing and tune the bus (`CMD6`)
12. Tune the SD clock if the timing is SDR50 or faster (`CMD19`)

The below diagram relates these steps to SDHC API calls. For more detail on the SD initialization flow, see SD specifications "Part 1: Physical Layer" and "Part A2 SD Host Controller Specification"

![Detailed SD init flow](https://user-images.githubusercontent.com/34665051/155623164-80cb5156-2998-431c-a13c-50a439016bbe.png)

danieldegrasse committed:
drivers: sdhc: Add SD host controller driver API

Add api for SD host controller driver. SD host controller driver
supports the following operations:

- reset: reset host controller state
- request: send SD command and data via SDHC
- set_io: set I/O settings (voltage, clocks, etc..) on SDHC
- get_card_present: check for card presence
- execute_tuning: run tuning process for UHS cards
- card_busy: check if SD card bus is busy
- get_host_props: get host controller properties

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: sdhc: Implement NXP USDHC SDHC driver

Implement SDHC driver for NXP USDHC peripheral, supporting all api calls
available in the sdhc driver. This implementation leverages NXP's HAL,
and simply implements a shim layer over the HAL itself.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: sdhc: Add SDHC DTS bindings

Add generic SDHC dts binding, as well as DTS binding for NXP USDHC.
Update iMX.RT DTS binding to use USDHC compatible

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
include: sd: Add SD specification definition

SD specification definition required for SD host controller headers.
Add header file with SD specification definitions.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1064: Enabled new SDHC driver

Enabled new SDHC driver for mimxrt1064 evk

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
tests: sdhc: Add SD host controller driver test

SD host controller driver runs basic SD host controller tests, including
checking SD presence, and sending commands to SD card. No data transfer
is performed.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
include: sd: Add SD subsystem header files

Add SD subsystem headers. SD subsystem contains generic header for SD
initialization, and headers for SDIO and SDMMC cards.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
sd: Add common SD initialization code

All SD cards require SD CMD0 (reset) and CMD8 (send IF cond) at boot.
Add this portion of the initialization flow to SD subsystem, as well as
query command to check if card is SDIO.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
sd: Add sdmmc protocol stack

Add SDMMC driver to subsystem. SDMMC driver will handle initialization,
as well as SDMMC I/O. SD mode support is currently supported, SPI mode
support is not.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
tests: subsys: sd: Add SDMMC subsystem test

Add SDMMC subsystem test, to verify functionality of SDMMC portion of
subsystem.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: disk: add SDMMC zephyr disk driver

Add generic sdmmc zephyr disk driver, which uses the SDMMC subsystem

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: Enable USDHC driver for all RT10xx based boards

Enable new USDHC driver for all RT10xx boards, since those will have
the SDHC driver selected by Kconfig

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang commented:
test PASS, on rt1064, and CI regression for other platform is running
```
*** Booting Zephyr OS build zephyr-v3.0.0-486-g208f225f79a0  ***
Running test suite sdhc_api_test
===================================================================
START - test_reset
 PASS - test_reset in 0.1 seconds
===================================================================
START - test_host_props
 PASS - test_host_props in 0.1 seconds
===================================================================
START - test_set_io
 PASS - test_set_io in 0.1 seconds
===================================================================
START - test_card_presence
 PASS - test_card_presence in 0.501 seconds
===================================================================
START - test_card_if_cond
Found SDHC/SDXC card
 PASS - test_card_if_cond in 1.3 seconds
===================================================================
Test suite sdhc_api_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
*** Booting Zephyr OS build zephyr-v3.0.0-486-g208f225f79a0  ***
Running test suite sdhc_api_test
===================================================================
START - test_reset
 PASS - test_reset in 0.1 seconds
===================================================================
START - test_host_props
 PASS - test_host_props in 0.1 seconds
===================================================================
START - test_set_io
 PASS - test_set_io in 0.1 seconds
===================================================================
START - test_card_presence
 PASS - test_card_presence in 0.501 seconds
===================================================================
START - test_card_if_cond
Found SDHC/SDXC card
 PASS - test_card_if_cond in 1.3 seconds
===================================================================
Test suite sdhc_api_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
*** Booting Zephyr OS build zephyr-v3.0.0-486-g208f225f79a0  ***
Running test suite sd_stack_test
===================================================================
START - test_init
I: Detected SD card
I: Card switched to 1.8V signaling
 PASS - test_init in 1.168 seconds
===================================================================
START - test_ioctl
SD card reports sector count of 61030400
SD card reports sector size of 512
 PASS - test_ioctl in 0.7 seconds
===================================================================
START - test_read
 PASS - test_read in 0.18 seconds
===================================================================
START - test_write
 PASS - test_write in 0.383 seconds
===================================================================
START - test_rw
 PASS - test_rw in 1.951 seconds
===================================================================
START - test_card_config
Card voltage: 1.8V
Card timing: SDR104
Card type: SDMMC
Card spec: 3.0
 PASS - test_card_config in 0.7 seconds
===================================================================
Test suite sd_stack_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL

```

hakehuang commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk.dts:
```diff
@@ -220,7 +220,6 @@ zephyr_udc0: &usb1 {
 
 &usdhc1 {
 	status = "okay";
-	no-1-8-v;
```
why remove this?

danieldegrasse replied:
We have been using the `no-1-8-v` flag to block the usdhc driver from switching cards to 1.8V signalling, because 1.8V signalling levels enable higher clock speeds. The USDHC driver we currently have in the tree typically would fail to transfer data from the SD card at these higher clock speeds, so we disabled using 1.8V altogether with this flag. However, the RT1050 does have the required circuitry to switch an SD card to 1.8V signalling levels, so with the new SDHC driver we can reenable this feature.

hakehuang commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -621,21 +621,29 @@
 		};
 
 		usdhc1: usdhc@402c0000 {
-			compatible = "nxp,imx-usdhc";
```
usdhc is the IP name listed in our reference manual, is it possible to keep to usdhc?

danieldegrasse replied:
That name would conflict with NXP's current USDHC driver. Once all platforms using that driver transition to the new SDHC driver, we could use the `nxp,imx-usdhc` binding for this new SDHC driver, but for the time being this name is required to allow both drivers to coexist.

jfischer-no replied:
Is USDHC NXP's own design or can it be found in the SoC of other vendors?

danieldegrasse replied:
USDHC is NXP's own design as far as I am aware. SD host controllers do have a specification they conform to, but I opted not to implement a generic SD host controller driver with vendor specific overrides (the approach Linux takes) for a few reasons:

1. Zephyr allows vendor HALs, so most vendor drivers can be implemented as a shim and rely on tested HAL code to manage the SDHC IP
2. Code size should be reduced by skipping the abstraction of a common SD host controller driver with vendor overrides
3. If NXP's USDHC is anything to go by, SD host controllers often deviate from the SD specification, so a generic layer may not be very useful.

MaureenHelm replied:
The old and new SDHC drivers won't coexist in the same application though; there are Kconfig symbols to enable one or the other. I don't think this compatible should change.

hakehuang commented on include/sd/sd.h:
```diff
@@ -0,0 +1,100 @@
+	enum card_status status; /*!< Card status */
+	enum card_type type; /*!< Card type */
+	uint32_t flags; /*!< Card flags */
+	uint8_t card_buffer[CONFIG_SD_BUFFER_SIZE]
```
does this buffer need NOCACHE?

danieldegrasse replied:
No, the only buffer than needs to be explicitly marked as noncacheable is the ADMA descriptor buffer present in the SD host controller driver. The USDHC HAL driver implements the required cache maintenance operations for transfers to be made from cacheable memory.

de-nordic commented:
@ngboonkhai You may also want to take a look at this.

danieldegrasse committed:
soc: rt11xx: Enable USDHC SD host controller on RT1170

Enable SD host controller driver for RT1170, so the EVK can use the new
SD subsystem.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

jfischer-no commented on drivers/sdhc/imx_usdhc.c:
```diff
@@ -0,0 +1,200 @@
+/*
```
Can you please squash all subsequent `drivers/sdhc/imx_usdhc.c` commits?

danieldegrasse replied:
Yes, do you think the same should be done for the subsystem commits? My objective with the current commit order was to show the implementation process for future reference, and to make review easier. However, I'm aware it's a large commit history, and doesn't really follow the current methodology for introducing new files.

danieldegrasse commented on include/drivers/sdhc.h:
```diff
@@ -0,0 +1,437 @@
+	uint32_t response[4]; /*!< SD card response field */
+	enum sdhc_rsp_type response_type; /*!< Expected SD response type */
+	unsigned int retries; /*!< Max number of retries */
+	int timeout_ms; /*!< Command timeout in milliseconds */
```
I would like to allow the user to pass a special "forever" value to the timeout, like `K_FOREVER` for kernel APIs. Would you suggest moving to `k_timeout_t` for this field in that case?

jfischer-no replied:
I think it is implementation dependent, probably yes.

jfischer-no replied:
I guess it would simplify things like

```c
	if (cmd->timeout_ms == SDHC_TIMEOUT_FOREVER) {
		request.command_timeout = K_FOREVER;
	} else {
		request.command_timeout = K_MSEC(cmd->timeout_ms);
	}
```

danieldegrasse replied:
Do you mean by adding a layer like this to the `sdhc.h` file, that checks the timeout set on the command and adds it to the request? The other approach I'm considering is having a default timeout, since it is rarely desirable to have a zero timeout command (usually the user wants a command with a short, but nonzero timeout). The user could then override the timeout only when desired.

jfischer-no commented on include/drivers/sdhc.h:
```diff
@@ -0,0 +1,437 @@
+
+static inline int z_impl_sdhc_hw_reset(const struct device *dev)
+{
+	const struct sdhc_driver_api *api =
+		(const struct sdhc_driver_api *)dev->api;
+	if (!api->reset) {
+		return -ENOSYS;
+	}
+	return api->reset(dev);
```
Empty line after variable definition, empty line after `if(){}` block or before `return`.

danieldegrasse replied:
Addressed in latest push

danieldegrasse replied:
Fixed in latest push

danieldegrasse committed:
drivers: sdhc: add SD SPI mode host controller driver

Add SDHC driver implementing spi mode support for SD cards. This driver
implements the standard SD host controller APIs, and sets the host
property "is_spi" to indicate to the SD subsystem the card will be
running in SPI mode.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso55s69_cpu0: enable spi mode SD card.

Enable SD card running in SPI mode, using the SD subsystem.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: imx_rt: added support for nxp imx_usdhc SDHC driver to RT600/500

added support for NXP iMX RT600/RT500 to use to SDHC driver, with SD
subsystem. Tested with RT685 EVK

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: disk: remove legacy nxp USDHC driver

all in tree SOCs with the USDHC peripheral have now been converted to
use the new SD host controller USDHC driver, so remove legacy NXP disk
USDHC driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented on include/drivers/sdhc.h:
```diff
@@ -0,0 +1,437 @@
+	uint32_t max_current_180; /*!< Max current (in mA) at 1.8V */
+};
+
+typedef int (*sdhc_api_reset_t)(const struct device *dev);
+typedef int (*sdhc_api_request_t)(const struct device *dev,
+				struct sdhc_command *cmd,
+				struct sdhc_data *data);
+typedef int (*sdhc_api_set_io_t)(const struct device *dev, struct sdhc_io *ios);
+typedef int (*sdhc_api_get_card_present_t)(const struct device *dev);
+typedef int (*sdhc_api_execute_tuning_t)(const struct device *dev);
+typedef int (*sdhc_api_card_busy_t)(const struct device *dev);
+typedef int (*sdhc_api_get_host_props_t)(const struct device *dev,
+					struct sdhc_host_props *props);
```
Addressed in latest push

danieldegrasse commented on include/sd/sd_spec.h:
```diff
@@ -0,0 +1,543 @@
+/* R1 response status */
+enum sdhc_r1_status {
+	/* Bits 0-2 reserved */
+	SDHC_R1_AUTH_ERR = (1U << 3),
```
Addressed in latest push

danieldegrasse commented:
This PR has been updated to include another SD host controller driver, which implements the SD over SPI protocol (`sdhc_spi.c`), and the sd subsystem itself has been updated to handle cards running in SPI mode. Since this implementation now has multiple SD host controller drivers, I feel it is ready for review and I'm removing the WIP tag.

danieldegrasse commented on drivers/sdhc/imx_usdhc.c:
```diff
@@ -0,0 +1,924 @@
+	/* Reset transfer status */
+	dev_data->transfer_status = 0U;
+	/* Reset semaphore */
+	k_sem_reset(&dev_data->transfer_sem);
```
Multiple threads accessing the SD host controller would be an issue, yeah. Currently mutexes ensuring that only one thread accesses the SD card are handled at the subsystem, not SD host controller level.

Do you think the overhead of another mutex here would be valuable? Generally speaking, an application should never directly interact with the SD host controller APIs

hakehuang replied:
multi-thread access is very like in rtos, and either we use a queue access or we can use mutex. Maybe this can be leaved to enhancement later

hakehuang reviewed (commented):
some commemnts

danieldegrasse commented on drivers/sdhc/imx_usdhc.c:
```diff
@@ -0,0 +1,924 @@
+	dev_data->transfer_status = 0U;
+	/* Reset semaphore */
+	k_sem_reset(&dev_data->transfer_sem);
+#ifdef CONFIG_IMX_USDHC_DMA_SUPPORT
```
Both transfer functions here are non-blocking, the difference is that one uses the ADMA engine in the USDHC IP block, and the other doesn't. Some of NXP's SOCs do not support noncacheable memory in Zephyr, so they cannot support using DMA for transfers.

The reason for a non blocking API call here is that the non blocking API uses interrupts. So while the api call will block the current thread, other threads can execute freely until the transfer completes.

danieldegrasse commented on boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts:
```diff
@@ -178,13 +178,19 @@ zephyr_udc0: &usb1 {
 &usdhc1 {
 	status = "okay";
-	cd-gpios = <&gpio1 23 GPIO_ACTIVE_LOW>;
 	no-1-8-v;
 	pinctrl-0 = <&pinmux_usdhc1>;
 	pinctrl-1 = <&pinmux_usdhc1_slow>;
 	pinctrl-2 = <&pinmux_usdhc1_med>;
 	pinctrl-3 = <&pinmux_usdhc1_fast>;
 	pinctrl-names = "default", "slow", "med", "fast";
+	cd-gpios = <&gpio3 19 GPIO_ACTIVE_LOW>;
```
yes, it was. the `SD_CD_SW` signal on the RT1020 EVK is connected to `GPIO_SD_B0_06`, which corresponds to GPIO3 pin 19

danieldegrasse commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -286,14 +286,18 @@ zephyr_udc0: &usb1 {
 	status = "okay";
-	no-1-8-v;
-	pwr-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>;
+	pwr-gpios = <&gpio1 19 GPIO_ACTIVE_HIGH>;
```
It was, yeah. `SD_PWREN` is routed to `GPIO_AD_B1_03` on the RT1064 EVK, which uses GPIO1 pin 19. I'm not certain why this wasn't an issue before, but when I was testing the SD stack on the RT1064 I noticed that power didn't seem to be toggled to the SD card during the initialization phase, which led me to look at this GPIO selection.

carlescufi commented:
API meeting:

- Remove `sdmmc_spi.c` 
- Remove `sdio.h`
- Add a note in the release notes for current users or the disk subsystem

danieldegrasse committed:
boards: use zephyr,sdhc-spi-slot over zephyr,mmc-spi-slot

switch all in tree usage of zephyr,mmc-spi-slot to zephyr,sdhc-spi-slot.
This will change all boards to use the new SD subsystem instead of the
SDMMC SPI driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: disk: remove legacy SDMMC SPI driver

remove existing SDMMC SPI driver, since it is replaced by the SPI mode
SD host controller driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
doc: peripherals: add documentation for SDHC peripheral

add documentation for SDHC peripheral driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
doc: storage: update storage documentation for SD subsystem

add section about SD subsystem to disk documentation, and document new
SDHC SPI bindings

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
doc: release: add release notes for SD subsystem

add release notes about SDHC driver, as well as binding changes for SPI
mode SD cards.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
> API meeting:
> 
>  * Remove `sdmmc_spi.c`
> 
>  * Remove `sdio.h`
> 
>  * Add a note in the release notes for current users or the disk subsystem
> 
>  * Remaining question about presence detection and presence detection notifications (i.e. callbacks) in the SDHC driver.

`sdmmc_spi.c` has been removed, as well as `sdio.h`. The release notes have been updated to reflect this change. Additonally, I have fixed the stack to properly support legacy SDSC cards, and microcontrollers using SD SPI mode that do not support the standard clock frequency of 25 MHz for data transfer.

@henrikbrixandersen could you run `tests/subsys/sd/sdmmc` on the board you have that uses SD over SPI, to verify that SPI support is working for you? I haven't encountered any issues with the lpcxpresso55s69

lairdjm commented:
Not working here on nRF52840-based board with SD card via SPI, using:
```
&spi1 {
        compatible = "nordic,nrf-spi";
        status = "okay";
        pinctrl-0 = <&spi1_default>;
        pinctrl-1 = <&spi1_sleep>;
        pinctrl-names = "default", "sleep";
        cs-gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
        sdhc0: sdhc@0 {
                compatible = "zephyr,sdhc-spi-slot";
                reg = <0>;
                label = "SDHC_0";
                max-bus-freq = <8000000>;
                mmc: mmc {
                        compatible = "zephyr,sdmmc-disk";
                        status = "okay";
                        label = "SDMMC_0";
                };
        };
};
```
in the dts, output:
```
*** Booting Zephyr OS build zephyr-v3.0.0-2488-gc24410b7ccad  ***
Running test suite sd_stack_test
===================================================================
START - test_init
E: Timeout waiting for transfer complete
E: Card SCLK init sequence failed
E: Could not disable card power via SDHC

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:39: test_init: (ret not equal to 0)
Card initialization failed
 FAIL - test_init in 1.224 seconds
===================================================================
START - test_ioctl
SD card reports sector count of 0
SD card reports sector size of 0
 PASS - test_ioctl in 0.7 seconds
===================================================================
START - test_read

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:66: test_read: (ret not equal to 0)
Single block card read failed
 FAIL - test_read in 0.13 seconds
===================================================================
START - test_write

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:102: test_write: (ret not equal to 0)
Single block card write failed
 FAIL - test_write in 0.14 seconds
===================================================================
START - test_rw

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:139: test_rw: (ret not equal to 0)
Write to card failed
 FAIL - test_rw in 0.16 seconds
===================================================================
START - test_card_config

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:196: test_card_config: (Reached unreachable code)
Card voltage is not known value
 FAIL - test_card_config in 0.15 seconds
===================================================================
Test suite sd_stack_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

lairdjm commented:
Might be something wrong with the setup for that board, will need to investigate that. Is working on nRF5340 board with the same MicroSD card over SPI:
```
*** Booting Zephyr OS build zephyr-v3.0.0-2488-gc24410b7ccad  ***
Running test suite sd_stack_test
===================================================================
START - test_init
I: Detected SD card
I: Maximum SD clock is under 25MHz, using clock of 8000000Hz
 PASS - test_init in 2.174 seconds
===================================================================
START - test_ioctl
SD card reports sector count of 62333952
SD card reports sector size of 512
 PASS - test_ioctl in 0.7 seconds
===================================================================
START - test_read
 PASS - test_read in 0.752 seconds
===================================================================
START - test_write
 PASS - test_write in 0.327 seconds
===================================================================
START - test_rw
 PASS - test_rw in 1.156 seconds
===================================================================
START - test_card_config
Card voltage: 3.3V
Card timing: SDR12
Card type: SDMMC
Card spec: 3.0
 PASS - test_card_config in 0.7 seconds
===================================================================
Test suite sd_stack_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Also tried it with some ancient 128MB MicroSD card that came with a devkit, it has never been detected in zephyr with the previous driver/code, seems to lock the test up:
```
*** Booting Zephyr OS build zephyr-v3.0.0-2488-gc24410b7ccad  ***
Running test suite sd_stack_test
===================================================================
START - test_init
I: Card does not support CMD8, assuming legacy card
I: Detected SD card
I: Card does not support CMD8, assuming legacy card
I: Maximum SD clock is under 25MHz, using clock of 8000000Hz
```
And results from an also equally old 512MB SD card:
```
*** Booting Zephyr OS build zephyr-v3.0.0-2488-gc24410b7ccad  ***
Running test suite sd_stack_test
===================================================================
START - test_init
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Legacy card detected, no CMD8 support
I: Card does not support CMD8, assuming legacy card
E: No usable card type was found
E: Unusable card

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:39: test_init: (ret not equal to 0)
Card initialization failed
 FAIL - test_init in 2.66 seconds
===================================================================
START - test_ioctl
SD card reports sector count of 0
SD card reports sector size of 0
 PASS - test_ioctl in 0.7 seconds
===================================================================
START - test_read

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:66: test_read: (ret not equal to 0)
Single block card read failed
 FAIL - test_read in 0.13 seconds
===================================================================
START - test_write

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:102: test_write: (ret not equal to 0)
Single block card write failed
 FAIL - test_write in 0.13 seconds
===================================================================
START - test_rw

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:139: test_rw: (ret not equal to 0)
Write to card failed
 FAIL - test_rw in 0.16 seconds
===================================================================
START - test_card_config
Card voltage: 3.3V

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/sd/sdmmc/src/main.c:198: test_card_config: (card.status not equal to CARD_IN)
Card status is not OK
 FAIL - test_card_config in 0.17 seconds
===================================================================
Test suite sd_stack_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

danieldegrasse commented:
> Also tried it with some ancient 128MB MicroSD card that came with a devkit, it has never been detected in zephyr with the previous driver/code

@lairdjm could you run the test with `CONFIG_SD_LOG_LEVEL_DBG=y`? I'd be interested to see where the test is failing. I've tried several legacy SDSC cards on the mimxrt685_evk, and haven't seen any issues.

lairdjm commented:
Output is:
```
*** Booting Zephyr OS build zephyr-v3.0.0-2488-g86a7e8aaf1e8  ***
Running test suite sd_stack_test
===================================================================
START - test_init
D: Resetting power to card
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
I: Card does not support CMD8, assuming legacy card
I: Detected SD card
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
D: SD CMD8 failed with error -22
I: Card does not support CMD8, assuming legacy card
D: CMD10 failed: -22
D: Card MID: 0x0, OID: 
D: CMD9 failed: -22
D: Card block count 0, block size 512
I: Maximum SD clock is under 25MHz, using clock of 8000000Hz
```
All I remember about this micro SD card is it came with an ST dev board and worked on that. It enumerates on a PC:
```
[12884.002191] mmc0: new SD card at address b368
[12884.012662] mmcblk0: mmc0:b368 ST128 121 MiB 
[12884.013966]  mmcblk0: p1
```

danieldegrasse commented:
@lairdjm the return code indicates an illegal command, based off the fact that CMD10 is only used in SPI mode (in SD mode it requires the card address, which the stack has not queried at this point), it's *possible* that your card is not correctly entering SPI mode. Could you ensure that you are using `cs-gpios` and not the SPI peripheral's chip select, to rule that possibility out?

Another possible issue is that your card has trouble with CRC calculations in SPI mode. To test this you would remove the call to `sd_enable_crc` on line 109 of `subsys/sd/sd.c`. If the card works without that call, then the CRC check is to blame.

Finally it is possible this card just doesn't behave when in SPI mode. Have you used this card on other devices in SPI mode?

lairdjm commented:
The overlay I am testing with is copied from the BL5340 one (also tested on the BL5340 board with the changes in this review), and it works with the other card:
```
&spi3 {
        compatible = "nordic,nrf-spim";
        status = "okay";
        cs-gpios = <&gpio1 12 GPIO_ACTIVE_LOW>;

        pinctrl-0 = <&spi3_default>;
        pinctrl-1 = <&spi3_sleep>;
        pinctrl-names = "default", "sleep";
        sdhc@0 {
                reg = <0>;
                compatible = "zephyr,sdhc-spi-slot";
                status = "okay";
                label = "SDHC_0";
                max-bus-freq = <8000000>;
                mmc {
                        compatible = "zephyr,sdmmc-disk";
                        status = "okay";
                        label = "SDMMC_0";
                };
        };
};
```
It's entirely possible that the SD card is non-standard, has incompatibility issues or lacks SPI mode. Removing the CRC check does not fix it. Just worry some with one card as it seems to lock the zephyr application up

MaureenHelm commented on CODEOWNERS:
```diff
@@ -610,6 +610,7 @@
 /include/shell/                           @jakub-uC @nordic-krch
 /include/shell/shell_mqtt.h               @ycsin
 /include/sw_isr_table.h                   @dcpleung @nashif @andyross
+/include/sd/                              @danieldegrasse
```
The header files need to be moved under `include/zephyr`

MaureenHelm commented on subsys/sd/sd.c:
```diff
@@ -97,7 +97,7 @@ static int sd_common_init(struct sd_card *card)
 	/* Perform voltage check using SD CMD8 */
 	ret = sd_retry(sd_send_interface_condition, card, CONFIG_SD_RETRY_COUNT);
-	if (ret == ETIMEDOUT) {
+	if (ret == -ETIMEDOUT) {
```
Rebasing mistake?

danieldegrasse replied:
Indeed, should be fixed now

MaureenHelm reviewed (changes requested):
Couple of nits, but otherwise looks great.

I took it for a spin with `samples/subsys/fs/fat_fs` and `mimxrt1050_evk` and it works well. I don't have hardware to try the SPI driver.

danieldegrasse committed:
drivers: imx_usdhc: change DT_COMPAT string to imx-usdhc

with the legacy USDHC driver fully removed from the tree, the
nxp,imx-usdhc binding can now be used for the new SD host controller
driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
doc: add sdhc api to API overview list

add SDHC api to the api overview list, as an experimental API introduced
in 3.1.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@MaureenHelm I updated this PR to move the header files, and address other requested changes. I changed the compatible string for the SDHC USDHC driver in a second commit (after removing the legacy disk USDHC driver), since using the same compatible initially would cause build failures in the commit series where only the RT1064 was enabled with the SDHC driver, and the rest of the RT series was not.

@lairdjm I can't reproduce your issue on any card I've tried, so I'm not sure if I found the issue, but I did identify a function `sdhc_skip` that was reading from the SD card without a timeout in the SPI driver. Could you test again and see if the card at least fail in a predictable manner?

lairdjm commented:
Tested it and it now fails gracefully!

nordicjm reviewed (commented):
Needs details adding to release notes in doc/releases

danieldegrasse commented:
> Needs details adding to release notes in doc/releases

@lairdjm I have added a section detailing the addition of the SD subsystem into Zephyr in the release notes. Does that look good to you?

jfischer-no commented on dts/bindings/sdhc/zephyr,sdhc-spi-slot.yaml:
```diff
@@ -0,0 +1,5 @@
+
+compatible: "zephyr,sdhc-spi-slot"
+
+include: [sdhc.yaml]
```
actually, spi device should `include: spi-device.yaml`

danieldegrasse replied:
Changed to `spi-device.yaml`.

MaureenHelm commented on CODEOWNERS:
```diff
@@ -614,6 +614,7 @@
 /include/zephyr/shell/                           @jakub-uC @nordic-krch
 /include/zephyr/shell/shell_mqtt.h               @ycsin
 /include/zephyr/sw_isr_table.h                   @dcpleung @nashif @andyross
+/include/sd/                                     @danieldegrasse
```
`/include/zephyr/sd`

MaureenHelm commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -0,0 +1,434 @@
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_SDHC_H_
+#define ZEPHYR_INCLUDE_DRIVERS_SDHC_H_
+#include <device.h>
+#include <sd/sd_spec.h>
```
```suggestion
#include <zephyr/device.h>
#include <zephyr/sd/sd_spec.h>
```

MaureenHelm commented on drivers/sdhc/imx_usdhc.c:
```diff
@@ -0,0 +1,908 @@
+
+#define DT_DRV_COMPAT nxp_imx_sdhc
+
+#include <zephyr.h>
+#include <devicetree.h>
+#include <drivers/sdhc.h>
+#include <sd/sd_spec.h>
+#include <drivers/clock_control.h>
+#include <drivers/gpio.h>
+#include <logging/log.h>
+#include <soc.h>
```
these need the zephyr namespace too

MaureenHelm commented on tests/drivers/sdhc/src/main.c:
```diff
@@ -0,0 +1,160 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <drivers/sdhc.h>
+#include <device.h>
```
add zephyr namespace

MaureenHelm commented on include/zephyr/sd/sd.h:
```diff
@@ -0,0 +1,97 @@
+#ifndef ZEPHYR_INCLUDE_SD_SD_H_
+#define ZEPHYR_INCLUDE_SD_SD_H_
+
+#include <device.h>
+#include <drivers/sdhc.h>
```
add zephyr namespace

MaureenHelm commented on include/zephyr/sd/sdmmc.h:
```diff
@@ -0,0 +1,74 @@
+#ifndef ZEPHYR_INCLUDE_SD_SDMMC_H_
+#define ZEPHYR_INCLUDE_SD_SDMMC_H_
+
+#include <device.h>
+#include <drivers/sdhc.h>
+#include <sd/sd.h>
```
ditto

MaureenHelm reviewed (changes requested):
There are a few places where you need to use includes with the zephyr namespace; I highlighted  some but not all of them.

danieldegrasse commented:
@MaureenHelm All of the `#include` directives added in this PR should now use the zephyr namespace, please let me know if I missed any

danieldegrasse commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -0,0 +1,434 @@
+ * driver, using the @ref sdhc_get_host_props api.
+ */
+struct sdhc_host_caps {
+	uint8_t timeout_clk_freq: 5;
```
yes, fixed these fields.
